### PR TITLE
Add a diary frequency option

### PR DESCRIPTION
**Pull request by jeremcs**
This PR is a draft to implement different diary frequencies.
This feature has been in the issue tracker for [quite some time](https://github.com/vimwiki/vimwiki/issues/265).

Basically, it provides two new fields for wikis : `diary_frequency` and `diary_start_week_monday`. They respectively default to `daily` and `1` (true).
```vim
let g:vimwiki_list = [
    \ {'path': '~/vimwiki/weekly', 
    \  'diary_frequency': 'weekly',
    \  'diary_start_week_monday': 1}
    \]
```

I initially had more ambitions for this PR, but I'll stick to adding a weekly period for now.

As the end user, when you open today's entry, you open this week's entry.
It is represented by the first day of the week, so for this week, the file is named `2020-05-18.<your_ext>`.
Whether it's a Monday or a Sunday is decided by the `diary_start_week_monday` flag.

Likewise, yesterday's diary is last week's diary and tomorrow is next week's.

As you can see, this is basically done by changing the `vimwiki#diary#diary_date_link` to accept a reference timestamp and how many days or weeks (according to this diary's settings) we shift.

What do you think ?

Also, thanks for the great plugin!

-----

Steps for submitting a pull request:

- [x] **ALL** pull requests should be made against the `dev` branch!
- [x] Take a look at [CONTRIBUTING.MD](https://github.com/vimwiki/vimwiki/blob/dev/CONTRIBUTING.md)
- [x] Reference any related issues.
- [x] Provide a description of the proposed changes.
- [x] PRs must pass Vint tests and add new Vader tests as applicable.
  No errors _added_. (I already have errors on dev when running locally. I don't know if it's due to my setup.)
- [x] Make sure to update the documentation in `doc/vimwiki.txt` if applicable, including the Changelog and Contributors sections.

### CODE REVIEW

**Comment by jeremcs** [comment]
In case someone's been reading this since yesterday: after more testing I understood that I just managed to dodge the many ways this PR could break. I fixed it and updated the description.

File: autoload/vimwiki/diary.vim [hunk 1]
```diff
@@ -33,10 +33,26 @@ endfunction
...
+
+  let l:delta = 0
+  if a:0 > 1
+    let l:delta = a:2
+  endif
+
+  let l:frequency = vimwiki#vars#get_wikilocal('diary_frequency')
+  if l:frequency == "weekly"
+    let l:start_week_monday = vimwiki#vars#get_wikilocal('diary_start_week_monday')
+    let l:computed_timestamp = l:timestamp + l:delta*60*60*24*7 - 60*60*24*((7-l:start_week_monday+str2nr(strftime("%w", l:timestamp))) % 7)
```
Comment (tinmarino) [inline autoload/vimwiki/diary.vim:49]: That is a long line, even a long equation, could you break it ? Better to take 5 lines : easyer to read and debug

File: autoload/vimwiki/diary.vim [hunk 2]
```diff
@@ -33,10 +33,26 @@ endfunction
...
   endif
+
+  let l:delta = 0
+  if a:0 > 1
+    let l:delta = a:2
+  endif
+
+  let l:frequency = vimwiki#vars#get_wikilocal('diary_frequency')
+  if l:frequency == "weekly"
+    let l:start_week_monday = vimwiki#vars#get_wikilocal('diary_start_week_monday')
```
Comment (tinmarino) [inline autoload/vimwiki/diary.vim:48]: You can start on monday or not on monday. Nice from you to add this user customization. Then almost for free you get it as a number like start_week_day from 1 to 7

File: plugin/vimwiki.vim [hunk 3]
```diff
@@ -363,11 +363,11 @@ command! -count=0 VimwikiTabMakeDiaryNote
 
 command! -count=0 VimwikiMakeYesterdayDiaryNote
       \ call vimwiki#diary#make_note(<count>, 0,
-      \ vimwiki#diary#diary_date_link(localtime() - 60*60*24))
+      \ vimwiki#diary#diary_date_link(localtime(), -1))
```
Comment (tinmarino) [inline plugin/vimwiki.vim:366]: Pretty cool : removes magics number everywhere. Even without the week feature, it is cleaner that way

**Comment by tinmarino** [comment]
Excelent job: as little changes as possible to get the feature.
Can be improve if you have time (1, 2h max). At least you deserve to author the doc of your feature (It'll be reviewed don't worry). Are you OK with this ?
Sorry for the late review: it is the fault of the word "Draft"

**Comment by jeremcs** [comment]
Thanks ! And no problem. I'll clean it up, write the doc and put it as ready for review in the next few days.

**Comment by jeremcs** [comment]
@tinmarino Documentation, changelog and monthly/yearly options are all done :+1: 
Date computations are a bit of a pain in vimscript, I tried to break down everything to improve readability.

**Comment by tinmarino** [comment]
- I waited to see if you added some change.
- The code is good for me: as better even with default behavior
- If The travis test failed due to the linter. I'll correct that in a subsequent commit:

```
autoload/vimwiki/diary.vim:61:18: Use robust operators `==#` or `==?` instead of `==` (see Google VimScript Style Guide (Matching))
autoload/vimwiki/diary.vim:61:21: Prefer single quoted strings (see Google VimScript Style Guide (Strings))
autoload/vimwiki/diary.vim:63:41: Prefer single quoted strings (see Google VimScript Style Guide (Strings))
autoload/vimwiki/diary.vim:69:22: Use robust operators `==#` or `==?` instead of `==` (see Google VimScript Style Guide (Matching))
autoload/vimwiki/diary.vim:69:25: Prefer single quoted strings (see Google VimScript Style Guide (Strings))
autoload/vimwiki/diary.vim:70:42: Prefer single quoted strings (see Google VimScript Style Guide (Strings))
autoload/vimwiki/diary.vim:74:51: Prefer single quoted strings (see Google VimScript Style Guide (Strings))
autoload/vimwiki/diary.vim:77:22: Use robust operators `==#` or `==?` instead of `==` (see Google VimScript Style Guide (Matching))
autoload/vimwiki/diary.vim:77:25: Prefer single quoted strings (see Google VimScript Style Guide (Strings))
autoload/vimwiki/diary.vim:78:41: Prefer single quoted strings (see Google VimScript Style Guide (Strings))
autoload/vimwiki/diary.vim:82:50: Prefer single quoted strings (see Google VimScript Style Guide (Strings))
```

**Pull request merged**